libvast/vast/data.hpp

```diff
@@ -356,6 +357,48 @@ struct sum_type_access<vast::data> : default_sum_type_access<vast::data> {};
 
 } // namespace caf
 
+namespace vast {
+
+// -- extraction -------------------------------------------------------
+
+/// Tries to retrieve the value associated to `name` from `xs`.
+const data* get_if(const record* xs, std::string_view name);
+
+/// Tries to retrieve the value associated to `name` from `xs`.
+template <class T>
+caf::expected<T> get_if(const record* xs, std::string_view name) {
+  auto data = get_if(xs, name);
+  if (!data)
+    return make_error(ec::convert_error, "record has no", std::string{name},
+                      "field");
+  auto value = caf::get_if<T>(data);
+  if (!value)
+    return make_error(ec::convert_error, std::string{name},
+                      "has an unexpected type:", *data);
+  return *value;
+}
+
+template <class T>
+caf::expected<std::vector<T>> get_vec(const record* xs, std::string_view name) {
```

Comment:
Instead of creating this one-off function, I have a separate idea:

1. Use `get_if` in combination with `list` as above
2. Write a utility function that checks that the value type of the list is a certain expected `T`.

Reply:
For example:

```cpp
auto xs = get_if<list>(&r, "xs");
if (xs && value_type<std::string>(*xs))
   // list is a sequence of strings
```

Not sure if `value_type` is a good function name, but you get the idea.

Reply:
It is unclear to me what `value_type` would actually do? From the example usage it seems to be
`std::all_of(xs->begin(), xs->end(), [&](data& x){ return caf::holds_alternative<T>(x);}`

TBH, I don't see how this would be useful.

Reply:
> TBH, I don't see how this would be useful.

But that's exactly what you do in this function, no?

Also, this is just one detail. The bigger picture is about `get_vec` as a whole.

Reply:
No, `get_vec` is a type conversion. It turns a turns a `list` of `data` containing `T` at the key `name` into a `std::vector<T>`. `value_type` otoh does only verification.

As a utility `get_vec` provides a lot of value, because it replaces multiple calls to `get_if` with one.

Reply:
I'd like to see documentation for this function and also a better name. There's clearly an expectation mismatch between what a function named `get_vec` should do and what the function actually does.